execute/README.md

```diff
@@ -0,0 +1,456 @@
+The data contained within a message may be retained and used in other ways.
+
+The following messages presently exist:
+
+|Name|Data|Description
+|----|----|-----------
+|ProcessChunk|table.Chunk|A table chunk is ready to be processed
+|FlushKey|group key|Data associated with the given group key should be flushed from the dataset
+|Finish|error (optional)|The upstream dataset will produce no more messages
+
+The following messages exist, but are deprecated and should not be used by future transformations.
+
+|Name|Data|Description
+|----|----|-----------
+|Process|flux.Table|A full table is ready to be processed
+|UpdateWatermark|time|Data older than the given time will not be sent
+|UpdateProcessingTime|time|Marks the present time
+|RetractTable|key|Data associated with the given group key should be retracted
+
+### Memory Types
+
+Data memory references memory used for storing user data.
+This is in contrast to process or program memory.
+Since flux is integrated in co-tenant environments, it needs to handle arbitrary user data and avoid crashing the system or causing a noisy neighbor performance problem.
+Flux does this by separating the concept of process memory and data memory.
+
+Process memory handles any memory required to execute code and allocate memory to support the execution of that code.
+Process memory is anything in Go that would allocate stack or heap memory and is tracked by the garbage collector.
+User code, by necessity, will use process memory.
+The amount of process memory used by a transformation should be designed to be fairly consistent.
+If a user has 10 rows or 10,000 rows, the process memory should not scale directly with the number of rows.
+This is not a strict requirement.
+There's no expectation that a transformation will use the same memory regardless of the number of rows, but just the requirement that it doesn't linearly scale with the number of rows or worse.
+
+Data memory has different conditions.
+Data memory is memory that is used to store user data and data memory is tracked by the `memory.Allocator`.
+The flux engine places limits on the amount of data memory that can be used and will abort execution if more is used.
+It is allowed and expected that some transformations will have bad memory footprints for certain inputs.
+
+The primary method of storing data is through immutable arrow arrays.
+There are also circumstances where mutable data is needed to implement an algorithm, but process memory is not an appropriate place to store that data (such as `distinct()` or `sort()`).
+See the **Mutable Data** section for how to handle these circumstances.
+
+## Execution Engine
+
+The execution engine is central to executing flux queries.
+
+### Pipeline Creation
+
+Before the execution engine begins, it's important to understand the steps that happen before execution.
+These steps will not be described in detail here.
```

Comment:
This is honestly a lot better than the current text. Makes the text come across as more helpful rather than the "good luck figure it out yourself sucker" message that the current text accidentally sounds like. I only really realized that when I saw this comment.